extension/src/format/sexpression.ts

```diff
@@ -890,4 +890,94 @@ export class Sexpression extends LispAtom {
 
         return this.format(startColumn);
     }
-} 
+
+
+    getSexpressionFromPos(loc: Position, parent: boolean = false): Sexpression {
+        let result: Sexpression = null;
+        if (this.contains(loc)) {
+            this.atoms.forEach(sexpr =>{
+                if (sexpr instanceof Sexpression && sexpr.contains(loc)) {
+                    result = sexpr.getSexpressionFromPos(loc, parent) ?? (parent ? this : sexpr);
+                }
+            });
+        }
+        return result;
+    }
+
+    getParentOfSexpression(child: Sexpression): Sexpression {        
+        const pos = new Position(child.line, child.column);
+        return this.getSexpressionFromPos(pos, true);
+    }
+
+    getRange(){
+        const begin: LispAtom = this.atoms[0];
+        const close: LispAtom = this.atoms[this.atoms.length -1];
+        return new Range(begin.line, begin.column, close.line, (close.column + close.symbol.length) - 1);
+    }
+
+    
+    contains(position: Position): boolean {
+        return this.getRange().contains(position);
+    }
+
+
+    // This version was necessary to properly alternate over SETQ Name vs Value 
+    private isValidForSetq(atom: LispAtom) {
+        return !atom.isComment() && !['\'', '(', ')', '.'].includes(atom.symbol) && (atom instanceof Sexpression || atom.symbol.trim().length > 0);
+    }    
+    // This is general purpose utility to make sure primitives such as strings, numbers and decorations are not evaluated
+    private isValidNonPrimitive(atom: LispAtom) {        
+        return !atom.isComment() && !['\'', '(', ')', '.'].includes(atom.symbol) && !(/^".*"$/.test(atom.symbol)) && !(/^\'{0,1}\-{0,1}\d+[eE\-]{0,2}\d*$/.test(atom.symbol));
+    }
+
+
+    // This is an ittoration helper that makes sure only useful LispAtom/Sexpression's get used for contextual data collection purposes.
+    nextKeyIndex(currentIndex: number, forSetq?: boolean): number {
+        let index = currentIndex + 1;
+        if (index < this.atoms.length) {
+            const atom = this.atoms[index];
+            const flag = forSetq ? this.isValidForSetq(atom) : this.isValidNonPrimitive(atom);
+            if (atom instanceof LispAtom && flag){
+                return index;               
+            } else {
+                return this.nextKeyIndex(index, forSetq);
+            }
+        } else {
+            return -1;
+        }
+    }
+
+    getNthKeyAtom(count: number){
```

Comment:
I am actually not sure what to call this variable. It's not an index and its not really a count either. Its closer to an index, but an index specific to the LispAtoms/Sexpressions that have actual significance.